src/modm/driver/encoder/as5600.hpp

```diff
@@ -0,0 +1,346 @@
+
+#pragma once
+
+#include <modm/architecture/interface/i2c_device.hpp>
+#include <modm/architecture/interface/register.hpp>
+
+#include "angle.hpp"
+
+namespace modm
+{
+
+struct as5600
+{
+
+	// @see datasheet page 19
+	enum Register : uint8_t
+	{
+		ZMCO = 0x00,
+		ZPOS = 0x01,       // two bytes
+		MPOS = 0x03,       // two bytes
+		MANG = 0x05,       // two bytes
+		CONF = 0x07,       // two bytes
+		ANGLE_RAW = 0x0C,  // two bytes
+		ANGLE = 0x0E,      // two bytes
+		STATUS = 0x0B,
+		AGC = 0x1A,
+		MAGNITUDE = 0x1B,  // two bytes
+		BURN = 0xFF,
+
+		// Only AS5600L supports changing the I2c address
+		I2C_ADDR = 0x20,
+		I2C_UPDT = 0x21,
+	};
+
+	// @see datasheet page 20
+	enum class Config : uint16_t
+	{
+		PM0 = Bit0,  // Power Mode
+		PM1 = Bit1,
+
+		HYST0 = Bit2,  // Hysteresis
+		HYST1 = Bit3,
+
+		OUTS0 = Bit4,  // Output Stage
+		OUTS1 = Bit5,
+
+		PWM0 = Bit6,  // PWM Frequency
+		PWM1 = Bit7,
+
+		SF0 = Bit8,  // Slow Filter
+		SF1 = Bit9,
+
+		FTH0 = Bit10,  // Fast Filter Threshold
+		FTH1 = Bit11,
+		FTH2 = Bit12,
+
+		WD = Bit13  // Watchdog    0: Off, 1: On
+	};
+	MODM_FLAGS16(Config);
+
+	enum class PowerMode : uint16_t
+	{
+		NOM = 0,
+		LPM1 = 1,
+		LPM2 = 2,
+		LPM3 = 3,
+	};
+	using PowerMode_t = Configuration<Config_t, PowerMode, 0b11, 0>;
+
+	enum class Hysteresis : uint16_t
+	{
+		OFF = 0,
+		LSB = 1,
+		LSB2 = 2,
+		LSB3 = 3
+	};
+	using Hysteresis_t = Configuration<Config_t, Hysteresis, 0b11, 2>;
+
+	enum class OutputStage : uint16_t
+	{
+		ANALOG = 0,
+		ANALOG_REDUCED = 1,
+		PWM = 2
+	};
+	using OutputStage_t = Configuration<Config_t, OutputStage, 0b11, 4>;
+
+	enum class PWMFrequency : uint16_t
+	{
+		_115Hz = 0,
+		_230_Hz = 1,
+		_460_Hz = 2,
+		_920_Hz = 3
+	};
+	using PWMFrequency_t = Configuration<Config_t, PWMFrequency, 0b11, 6>;
+
+	enum class SlowFilter : uint16_t
+	{
+		x16 = 0,
+		x8 = 1,
+		x4 = 2,
+		x2 = 3
+	};
+	using SlowFilter_t = Configuration<Config_t, SlowFilter, 0b11, 8>;
+
+	enum class FastFilterThreshold : uint16_t
+	{
+		_6LSB = 0,
+		_7LSB = 1,
+		_9LSB = 2,
+		_18LSB = 3,
+		_21LSB = 4,
+		_24LSB = 5,
+		_10LSB = 6
+	};
+	using FastFilterThreshold_t = Configuration<Config_t, FastFilterThreshold, 0b11, 10>;
+
+	enum class Burn : uint8_t
+	{
+		ANGLE = Bit3,  // Burn Angle data. CAN ONLY BE DONE ONCE!
+		CONFIG = Bit2  // Burn Config data. CAN ONLY BE DONE ONCE!
+	};
+
+	// @see datasheet page 21
+	enum class Status : uint8_t
+	{
+		MagnetTooStrong = Bit3,
+		MagnetTooWeak = Bit4,
+		MagnetDetected = Bit5
+	};
+	MODM_FLAGS8(Status);
+
+	using Data = modm::Angle<12>;
+};
+
+template<class I2cMaster>
+class As5600 : public as5600, public modm::I2cDevice<I2cMaster, 2>
+{
+public:
+	// AS5600 has hardwired address 0x36
+	// AS4500L has default address 0x40 but supports programming a different one
+	As5600(Data &data, uint8_t address = 0x36) : I2cDevice<I2cMaster, 2>(address), data(data) {}
+
+	/* Reset to Power up state.
```

Comment:
```suggestion
	/** Reset to Power up state.
```
for doxygen/doxypress.